File: capstone/week1/mock_interview/practice/minimum_size_subarray_sum.go
package main

func minSubArrayLen(target int, nums []int) int {
	a, r := 0, 0
	max := len(nums) + 1
	currentSum := 0
	for r < len(nums) {
		currentSum += nums[r]
		for currentSum >= target {
			curLen := r - a + 1
			if curLen == 1 {
				return 1
			}
			if curLen < max {
				max = curLen
			}
			currentSum -= nums[a]
			a++
		}
		r++
	}
	if max == len(nums)+1 {
		return 0
	}
	return max
}

File: capstone/week1/mock_interview/practice/minimum_size_subarray_sum.js
"use strict"

function minSubArrayLen(target, nums) {
  let minimumSize;

  for (let anchor = 0; anchor < nums.length; anchor += 1) {
    let runner = anchor;
    let sum = 0;

    while (runner < nums.length && sum < target) {
      sum += nums[runner];
      runner += 1;
    }

    if (sum === target && (minimumSize === undefined || minimumSize > runner - anchor)) {
      minimumSize = runner - anchor;
    }
  }

  if (minimumSize === undefined) {
    return 0;
  }

  return minimumSize;
}

console.log(minSubArrayLen(7, [2, 3, 1, 2, 4, 3]));
console.log(minSubArrayLen(4, [1, 4, 4]));
console.log(minSubArrayLen(11, [1, 1, 1, 1, 1, 1, 1, 1]));